pkg/vfs/vfsswift/impl.go

```diff
@@ -0,0 +1,352 @@
+package vfsswift
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/cozy/cozy-stack/pkg/vfs"
+	"github.com/ncw/swift"
+)
+
+var conn *swift.Connection
+
+type swiftVFS struct {
+	vfs.Indexer
+	c      *swift.Connection
+	domain string
+}
+
+// InitConnection should be used to initialize the connection to the
+// OpenStack Swift server.
+//
+// This function is not thread-safe.
+func InitConnection(fsURL *url.URL) error {
+	var err error
+	q := fsURL.Query()
+
+	var authURL *url.URL
+	auth := confOrEnv(q.Get("AuthURL"))
+	if auth == "" {
+		authURL = &url.URL{
+			Scheme: "http",
+			Host:   fsURL.Host,
+			Path:   "/identity/v3",
+		}
+	} else {
+		authURL, err = url.Parse(auth)
+		if err != nil {
+			return fmt.Errorf("vfsswift: could not parse AuthURL %s", err)
+		}
+	}
+
+	var username, password string
+	if q.Get("UserName") != "" {
+		username = confOrEnv(q.Get("UserName"))
+		password = confOrEnv(q.Get("Password"))
+	} else {
+		password = confOrEnv(q.Get("Token"))
+	}
+
+	conn = &swift.Connection{
+		UserName:       username,
+		ApiKey:         password,
+		AuthUrl:        authURL.String(),
+		Domain:         confOrEnv(q.Get("UserDomainName")),
+		Tenant:         confOrEnv(q.Get("ProjectName")),
+		TenantId:       confOrEnv(q.Get("ProjectID")),
+		TenantDomain:   confOrEnv(q.Get("ProjectDomain")),
+		TenantDomainId: confOrEnv(q.Get("ProjectDomainID")),
+	}
+	if err := conn.Authenticate(); err != nil {
+		log.Errorf("[vfsswift] Authentication failed with the OpenStack Swift server on %s",
+			authURL.String())
+		return err
+	}
+	return nil
+}
+
+// New returns a vfs.VFS instance associated with the specified indexer and the
+// swift storage url.
+func New(index vfs.Indexer, domain string) (vfs.VFS, error) {
+	if conn == nil {
+		return nil, errors.New("vfsswift: global connection is not initialized")
+	}
+	if domain == "" {
+		return nil, fmt.Errorf("vfsswift: specified domain is empty")
+	}
+	return &swiftVFS{
+		Indexer: index,
+		c:       conn,
+		domain:  domain,
+	}, nil
+}
+
+func confOrEnv(val string) string {
```

Comment:
Yes, the idea is to reuse our URL notation in our configuration file, but allowing to specify env variables. For instance:

```yaml

fs:
  # file system url - flags: --fs-url
  # default url is the directory relative to the binary: ./storage

  url: >
    swift://openstack/
    ?UserName=$OS_USERNAME
    &Password=$OS_PASSWORD
    &ProjectName=$OS_PROJECT_NAME
    &UserDomainName=$OS_USER_DOMAIN_NAME
```